File: cpp/interview/classic/lru_cache/lru_cache.cpp
// 手写 LRU Cache（双向链表 + 哈希）
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

class LRUCache {
public:
    explicit LRUCache(int capacity) : cap(capacity) {
        head = new Node();
        tail = new Node();
        head->next = tail;
        tail->prev = head;
    }

    ~LRUCache() {
        Node* cur = head;
        while (cur) {
            Node* nxt = cur->next;
            delete cur;
            cur = nxt;
        }
    }

    int get(int key) {
        auto it = idx.find(key);
        if (it == idx.end()) return -1;
        moveToFront(it->second);
        return it->second->val;
    }

    void put(int key, int value) {
        auto it = idx.find(key);
        if (it != idx.end()) {
            it->second->val = value;
            moveToFront(it->second);
            return;
        }
        Node* node = new Node{key, value, nullptr, nullptr};
        idx[key] = node;
        addToFront(node);
        if ((int)idx.size() > cap) {
            Node* evict = tail->prev;
            removeNode(evict);
            idx.erase(evict->key);
            delete evict;
        }
    }

    int size() const { return (int)idx.size(); }

private:
    struct Node {
        int key{0};
        int val{0};
        Node* prev{nullptr};
        Node* next{nullptr};
    };

    int cap;
    Node* head;
    Node* tail;
    unordered_map<int, Node*> idx;

    void addToFront(Node* node) {
        node->prev = head;
        node->next = head->next;
        head->next->prev = node;
        head->next = node;
    }

    void removeNode(Node* node) {
        node->prev->next = node->next;
        node->next->prev = node->prev;
        node->prev = node->next = nullptr;
    }

    void moveToFront(Node* node) {
        removeNode(node);
        addToFront(node);
    }
};

int main() {
    LRUCache c(2);
    c.put(1, 1);
    c.put(2, 2);
    assert(c.get(1) == 1);
    c.put(3, 3);
    assert(c.get(2) == -1);
    c.put(4, 4);
    assert(c.get(1) == -1);
    assert(c.get(3) == 3);
    assert(c.get(4) == 4);
    c.put(4, 40);
    assert(c.get(4) == 40);
    assert(c.size() == 2);
    cout << "lru_cache OK" << endl;
    return 0;
}

File: python/interview/classic/lru_cache/lru_cache.py
"""手写 LRU Cache：双向链表 + 哈希表。"""

from __future__ import annotations

from typing import Optional


class _Node:
    __slots__ = ("key", "val", "prev", "next")

    def __init__(
        self,
        key: int = 0,
        val: int = 0,
        prev: Optional["_Node"] = None,
        next: Optional["_Node"] = None,
    ) -> None:
        self.key = key
        self.val = val
        self.prev = prev
        self.next = next


class LRUCache:
    def __init__(self, capacity: int) -> None:
        if capacity <= 0:
            raise ValueError("capacity must be positive")
        self._cap = capacity
        self._map: dict[int, _Node] = {}
        self._head = _Node()
        self._tail = _Node()
        self._head.next = self._tail
        self._tail.prev = self._head

    def __len__(self) -> int:
        return len(self._map)

    def get(self, key: int) -> int:
        node = self._map.get(key)
        if node is None:
            return -1
        self._move_to_front(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        node = self._map.get(key)
        if node is not None:
            node.val = value
            self._move_to_front(node)
            return
        node = _Node(key, value)
        self._map[key] = node
        self._add_to_front(node)
        if len(self._map) > self._cap:
            evict = self._tail.prev
            assert evict is not None and evict is not self._head
            self._remove(evict)
            del self._map[evict.key]

    def _add_to_front(self, node: _Node) -> None:
        nxt = self._head.next
        assert nxt is not None
        node.prev = self._head
        node.next = nxt
        self._head.next = node
        nxt.prev = node

    def _remove(self, node: _Node) -> None:
        p, n = node.prev, node.next
        assert p is not None and n is not None
        p.next = n
        n.prev = p
        node.prev = node.next = None

    def _move_to_front(self, node: _Node) -> None:
        self._remove(node)
        self._add_to_front(node)


if __name__ == "__main__":
    c = LRUCache(2)
    c.put(1, 1)
    c.put(2, 2)
    assert c.get(1) == 1
    c.put(3, 3)
    assert c.get(2) == -1
    c.put(4, 4)
    assert c.get(1) == -1
    assert c.get(3) == 3
    assert c.get(4) == 4
    c.put(4, 40)
    assert c.get(4) == 40
    assert len(c) == 2
    print("lru_cache OK")
